openlabcmd/openlabcmd/zk.py

```diff
@@ -206,37 +207,114 @@ def delete_node(self, node_name):
+                raise exceptions.ValidationError("node_name_filter should be "
+                                                 "a list or string.")
+        if node_role_filter:
+            if isinstance(node_role_filter, str):
+                node_role_filter = [node_role_filter]
+            if not isinstance(node_role_filter, list):
+                raise exceptions.ValidationError("node_role_filter should be "
+                                                 "a list or string.")
+
         result = []
         for exist_node in self.list_nodes():
+            if node_name_filter and exist_node['name'] not in node_name_filter:
+                continue
+            if node_role_filter and exist_node['role'] not in node_role_filter:
+                continue
             path = '/ha/%s/%s_service' % (exist_node['name'],
                                           exist_node['role'])
             for service_name in self.client.get_children(path):
                 service_path = path + '/' + service_name
                 service_info = self.client.get(service_path)
+                mtime = service_info[1].mtime
                 service_info = self._bytesToDict(service_info[0])
+                service_info['updated_at'] = datetime.datetime.fromtimestamp(
+                    mtime / 1000).isoformat()
+                service_info['node_name'] = exist_node['name']
                 result.append(service_info)
         return sorted(result)
 
-    @_client_check_wrapper
-    def get_service(self, service_name, role=None):
-        if not role:
-            if (service_name in
-                    service.service_mapping['master_service']['necessary'] +
-                    service.service_mapping['master_service']['unnecessary']):
-                role = 'master'
-            elif (service_name in
-                  service.service_mapping['slave_service']['necessary'] +
-                  service.service_mapping['slave_service']['unnecessary']):
-                role = 'slave'
-            else:
-                exceptions.ClientError("Can't find service %s" % service_name)
+    def _get_service_role(self, service_name):
+        if (service_name in
+                service.service_mapping['master_service']['necessary'] +
+                service.service_mapping['master_service']['unnecessary']):
+            return 'master'
+        elif (service_name in
+              service.service_mapping['slave_service']['necessary'] +
+              service.service_mapping['slave_service']['unnecessary']):
+            return 'slave'
 
+        exceptions.ClientError("Can't find service %s" % service_name)
+
+    def _get_service_path_and_node(self, service_name, role):
         for exist_node in self.list_nodes():
             if exist_node['role'] == role:
                 path = '/ha/%s/%s_service/%s' % (exist_node['name'], role,
                                                  service_name)
-                result = self.client.get(path)
```

Comment:
Why do not check the service existed? if service_name=`not_existed_service`?

Reply:
This function is only used for generate a path which will be used later, if the service doesn't exist, it'll raise NotFound error later.

Reply:
OK, thanks for your clarification